src/platforms/ios/iostunnel.swift

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: MIT
+// Copyright © 2018-2020 WireGuard LLC. All Rights Reserved.
+
+import Foundation
+import NetworkExtension
+import os
+
+class PacketTunnelProvider: NEPacketTunnelProvider {
```

Comment:
All good! I wonder if we can keep the PacketTunnelProvider implementation and "extend" it. 

For instance we could add to the Info.plist these:
```
<key>NSExtension</key>
        <dict>
                <key>NSExtensionPointIdentifier</key>
                <string>com.apple.networkextension.packet-tunnel</string>
                <key>NSExtensionPrincipalClass</key>
                <string>$(PRODUCT_MODULE_NAME).MozillaVPNPacketTunnelProvider</string>
        </dict>
```

Then implement the MozillaVPNPacketTunnelProvider in this way:

```
class MozillaVPNPacketTunnelProvider : PacketTunnelProvider {
 ...
  override func handleAppMessage(_ messageData: Data, completionHandler: ((Data?) -> Void)? = nil) {
  ...
```

Reply:
That was my first instinct as well. I tried that but since the `adapter` property is private the subclass cannot access it and therefore would not work.